### PR TITLE
add trackVerifiedToken()

### PR DESCRIPTION
**Pull request by nickpatrick**

### CODE REVIEW

File: Example/Example/AppDelegate.swift [hunk 1]
```diff
@@ -22,9 +22,14 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
         self.requestLocationPermissions()
 
         // Replace with a valid test publishable key
-        Radar.initialize(publishableKey: "prj_test_pk_0000000000000000000000000000000000000000")
+        Radar.initialize(publishableKey: "org_test_pk_5857c63d9c1565175db8b00750808a66a002acb8")
```
Comment (nickpatrick) [inline Example/Example/AppDelegate.swift:25]: Will revert this

File: RadarSDK/RadarUtils.m [hunk 2]
```diff
@@ -156,6 +158,57 @@ + (NSString *)dictionaryToJson:(NSDictionary *)dict {
     } 
 }
 
++ (BOOL)isJailbroken {
```
Comment (nickpatrick) [inline RadarSDK/RadarUtils.m:161]: May revert this

File: RadarSDK/RadarAPIClient.m [hunk 3]
```diff
@@ -327,10 +331,21 @@ - (void)trackWithLocation:(CLLocation *_Nonnull)location
...
 
+                        if (encrypted) {
+                            if (!tokenObj) {
+                                return completionHandler(status, nil, nil, nil, nil, nil, nil);
+                            }
+
+                            NSString *token = (NSString *)tokenObj;
+
+                            return completionHandler(status, nil, nil, nil, nil, nil, token);
+                        }
```
Comment (lmeier) [inline RadarSDK/RadarAPIClient.m:347]: This codepath is completely forgoing updating the event & location update handlers, trips, & place updating.

As such, `trackVerified` and `trackVerifiedEncrypted` behave significantly differently -- wonder how we should make it clear to users the difference in behavior and effects (could see myself getting confused by this).

File: RadarSDK/Include/Radar.h [hunk 4]
```diff
@@ -185,6 +185,15 @@ typedef void (^_Nullable RadarBeaconCompletionHandler)(RadarStatus status, NSArr
  */
 typedef void (^_Nullable RadarTrackCompletionHandler)(RadarStatus status, CLLocation *_Nullable location, NSArray<RadarEvent *> *_Nullable events, RadarUser *_Nullable user);
 
+/**
+ Called when an encrypted track request succeeds, fails, or times out.
+
+ Receives the request status and, if successful, a JSON Web Token (JWT) containing an array of the events generated and the user. Decrypt the JWT server-side using your private key.
+
+ @see https://radar.com/documentation/sdk/ios
```
Comment (jsani-radar) [inline RadarSDK/Include/Radar.h:193]: should we be pointing to the iOS docs or fraud docs? comments above the callback and method are inconsistent at the moment

**Review by jsani-radar — COMMENTED** [review]
looking good overall - few small things

File: RadarSDK/RadarUtils.m [hunk 5]
```diff
@@ -156,6 +158,57 @@ + (NSString *)dictionaryToJson:(NSDictionary *)dict {
...
+    
+    NSError *error;
+    [data writeToFile:testPath options:NSDataWritingAtomic error:&error];
+    
+    // If no error occurred, the device is likely jailbroken
+    if (!error) {
+        NSLog(@"failed check 2");
+        [[NSFileManager defaultManager] removeItemAtPath:testPath error:nil];
+        return YES;
+    }
```
Comment (lmeier) [inline RadarSDK/RadarUtils.m:189]: The first test makes sense to me.

For this test here, I'd love to read up on why this test works — is there a SO or other resource you can point me to?

File: Example/Example/AppDelegate.swift [hunk 6]
```diff
@@ -22,9 +22,14 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
...
         Radar.setDelegate(self)
+        
+        Radar.trackVerified { status, location, events, user in
+
```
[the remark this answers is not in the document]
Comment (tderouin) [reply Example/Example/AppDelegate.swift:29]: I ran into a weird issue with this:

```
        
        Radar.trackVerifiedEncrypted{ ( status, token ) in
            NSLog("Got token:")
            NSLog(token!)
        }
```

The `token` printed was truncated to 1,024 characters, but the token returned from the server was 7,000+ characters. 

While running the debugger, and clicking `Copy`, would sometimes copy the truncated value, but not always:

image

I had to run `print (void)CFShow(token)` in the debugging console to show the entire token to confirm it was being captured.

[the remark this answers is not in the document]
Comment (nickpatrick) [reply Example/Example/AppDelegate.swift]: I ran into a similar issue testing on Android where `logcat` lines seem to be truncated to `4000` chars, but my token was something like `4300`. I just printed it in two parts. In any case, I reverted this. We can revisit inclusion in sample app when we revisit sample app code more broadly

File: RadarSDK/RadarAPIClient.m (hunk 3 above)
Comment (nickpatrick) [reply RadarSDK/RadarAPIClient.m:347]: Will mention on /documentation. Most people using this aren't going to be using `startTracking()`, so shouldn't cause unexpected behavior in practice

File: RadarSDK/Include/Radar.h (hunk 4 above)
Comment (nickpatrick) [reply RadarSDK/Include/Radar.h:193]: Good catch, will update

File: RadarSDK/Radar.m [hunk 7]
```diff
@@ -219,7 +219,7 @@ + (void)trackOnceWithLocation:(CLLocation *)location completionHandler:(RadarTra
                                               replayed:NO
                                                beacons:nil
                                      completionHandler:^(RadarStatus status, NSDictionary *_Nullable res, NSArray<RadarEvent *> *_Nullable events, RadarUser *_Nullable user,
-                                                         NSArray<RadarGeofence *> *_Nullable nearbyGeofences, RadarConfig *_Nullable config) {
+                                                         NSArray<RadarGeofence *> *_Nullable nearbyGeofences, RadarConfig *_Nullable config, NSString *_Nullable token) {
```
[the remark this answers is not in the document]
Comment (nickpatrick) [reply RadarSDK/Radar.m:222]: This callback is only ever used internally, not worried about

File: RadarSDK/RadarUtils.m (hunk 5 above)
Comment (nickpatrick) [reply RadarSDK/RadarUtils.m:189]: This doesn't actually work, reverting

File: RadarSDK/RadarUtils.m (hunk 2 above)
Comment (nickpatrick) [reply RadarSDK/RadarUtils.m]: Reverted

File: Example/Example/AppDelegate.swift (hunk 1 above)
Comment (nickpatrick) [reply Example/Example/AppDelegate.swift]: Reverted

File: RadarSDK/RadarAPIClient.h [hunk 8]
```diff
@@ -86,6 +87,7 @@ typedef void (^_Nonnull RadarSyncLogsAPICompletionHandler)(RadarStatus status);
         attestationString:(NSString *_Nullable)attestationString
                     keyId:(NSString *_Nullable)keyId
          attestationError:(NSString *_Nullable)attestationError
+                encrypted:(BOOL)encrypted
```
Comment (lmeier) [inline RadarSDK/RadarAPIClient.h:90]: was going to suggest we think about changing this to `verified` given discussion in slack & recent changes, but of course that's already used.. 🤔

Comment (lmeier) [reply RadarSDK/RadarAPIClient.h:90]: Looking at the JWT docs, I think `encoded` would perhaps be most accurate

Comment (nickpatrick) [reply RadarSDK/RadarAPIClient.h:90]: `token` param or `POST /track/token` endpoint are both options, too

**Pull request merged**